tensorflow_probability/python/distributions/empirical.py

```diff
@@ -0,0 +1,387 @@
+  dist.prob([0., 1.])
+  ==> 0.5
+  dist.prob([[0., 1.], [1., 2.]])
+  ==> [0.5, 0.25] # samples will be broadcast to shape [2, 4, 2] to match event.
+
+  # Initialize a empirical distribution with a [2] batch of vector samples.
+  dist = VectorEmpirical(samples=[[[0., 0.], [0., 1.]], [[0., 1.], [0., 1.]]])
+  dist.prob([0., 1.])
+  ==> [0.5, 1.] # event will be broadcast to shape [[0., 1.], [0., 1.]] to match samples.
+  ```
+
+  """
+
+  def __init__(self,
+               samples,
+               validate_args=False,
+               allow_nan_stats=True,
+               name="VectorEmpirical"):
+    """Initialize VectorEmpirical distributions.
+
+    Args:
+      samples: Numeric `Tensor` of shape [B1, ..., Bk, S, E]`, `k >= 0`.
+        Samples or batches of samples on which the distribution is based.
+        The first `k` dimensions index into a batch of independent distributions
+        and the last 2 dimension represents a multiset of vector-like samples
+        for each distribution.
+      validate_args: Python `bool`, default `False`. When `True` distribution
+        parameters are checked for validity despite possibly degrading runtime
+        performance. When `False` invalid inputs may silently render incorrect
+        outputs.
+      allow_nan_stats: Python `bool`, default `True`. When `True`, statistics
+        (e.g., mean, mode, variance) use the value "`NaN`" to indicate the
+        result is undefined. When `False`, an exception is raised if one or
+        more of the statistic's batch members are undefined.
+      name: Python `str` name prefixed to Ops created by this class.
+
+    Raises:
+      ValueError: if `samples` is a scalar.
+    """
+    super(VectorEmpirical, self).__init__(
+        samples,
+        event_ndims=1,
+        validate_args=validate_args,
+        allow_nan_stats=allow_nan_stats,
+        name=name)
+
+  def _mode(self):
+    def _get_mode(samples):
+      # TODO: Swith to tf.unique_with_counts_v2 when updated in TensorFlow API
+      y, idx, count = gen_array_ops.unique_with_counts_v2(samples, axis=[0])
+      mode = tf.gather(samples, tf.argmax(count))
+      return mode
+
+    event_size = tf.reshape(self.event_shape_tensor(), [])
+    samples = tf.reshape(self.samples, [-1, self.num_samples, event_size])
+    modes = tf.map_fn(_get_mode, samples)
+    modes_shape = tf.concat(
+        [self.batch_shape_tensor(), self.event_shape_tensor()],
+        axis=0)
+    return tf.reshape(modes, modes_shape)
+
+  def _sample_n(self, n, seed=None):
+    def _get_sample(samples):
+      indices = tf.random.uniform([n], maxval=self.num_samples,
+                                  dtype=tf.int32, seed=seed)
+      tensor = tf.gather(samples, indices, axis=-2)
+      return tensor
+
+    event_size = tf.reshape(self.event_shape_tensor(), [])
+    samples = tf.reshape(self.samples, [-1, self.num_samples, event_size])
+    draws = tf.map_fn(_get_sample, samples)
```

Comment:
I think you're doing more work here than you have to.
```python
event_ndims=1
x=tf.random.normal([2, 3, 4, 5, 6])
i=tf.random.uniform([10], maxval=x.shape[-event_ndims-1], dtype=tf.int32)

z=tf.gather(x, tf.random.uniform([10], maxval=x.shape[-event_ndims-1], dtype=tf.int32), axis=-event_ndims-1)
print(z.shape)
axes=tf.concat([[tf.rank(z)-event_ndims-1], tf.range(tf.rank(z)-event_ndims-1,dtype=tf.int32), tf.range(event_ndims) + tf.rank(z)-event_ndims], axis=0)
print(axes)
samples=tf.transpose(z, axes)
```
Notably, you don't have to collapse the event or the batch dims in this case.

Reply:
Furthermore, if you put this on the base class, I think it works generally for all event_ndims.